File: longest-substring-without-repeating-characters.py
from typing import List

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        chars = set()
        left = 0
        ans = 0

        for right, char in enumerate(s):
            while char in chars:
                chars.remove(s[left])
                left += 1
            chars.add(char)
            ans = max(ans, right-left+1)
        return ans

File: longest-substring-without-repeating-characters.cpp
#include <string>
#include <unordered_set>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
		unordered_set<char> chars;
		int ans = 0;
		int lt = 0;

		for (int rt=0; rt < s.size(); ++rt) {
			char ch = s[rt];
			while (chars.find(ch) != chars.end()) {
				chars.erase(s[lt]);
				++lt;
			}
			chars.insert(ch);
			ans = max(ans, rt-lt+1);
		}

		return ans;
    }
};
